Add ToTokens for Fields<syn::Field>

Luro02 opened the pull request:
closes #84 

The `Span` is kinda hacky and I know, that this might not work, but this would be a solution, that doesn't require a version bump.

The better solution would be to add a field `span: Span` to `Fields`, that contains the original `Span` from `syn::Fields`.

TedDriggs commented:
Can you put up a PR with the added-field fix? I agree that's probably the better solution here, and I don't mind bumping the minor version as long as it won't break existing uses.

Luro02 commented:
I fixed the fields and squashed the commits.

The problem with having all fields public is, that people could do this:
```rust
let fields = Fields {
    style: Style::Struct,
    fields: vec![],
}
```
and by adding a new field to the struct the above code would become invalid and everyone, who does this would have to change their code.

I made the `span` field intentionally private to prevent such problems in the future, this can be removed as soon as darling is close to 1.0. To accommodate for this change, I added a `new` function and a `with_span` function (inspired by `darling::Error`).

I had to implement `PartialEq` manually, because `Span`s do not implement it.
This does not seem to be a problem, because Spans can be ignored? At least, that is the case in syn with for example `Token`s, where `PartialEq` always returns true:
https://github.com/dtolnay/syn/blob/d42efb61a98bec129fbdb3bf6a79fa67b31f35aa/src/token.rs#L473

Luro02 commented:
The last commit changes
```rust
    /// Adds a [`Span`] to [`Fields`].
    pub fn with_span(mut self, span: Span) -> Self {
        self.span = Some(span);
        self
    }
```

to
```rust
    /// Adds a [`Span`] to [`Fields`].
    pub fn with_span(mut self, span: Span) -> Self {
        if self.span.is_none() {
            self.span = Some(span);
        }
        self
    }
```

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -150,17 +152,48 @@ impl<V: UsesLifetimes, F: UsesLifetimes> UsesLifetimes for Data<V, F> {
 pub struct Fields<T> {
     pub style: Style,
     pub fields: Vec<T>,
+    // This field is intentionally private, to make it easier to add new fields to this struct
```
Rather than making this field private, how about we add it as public and add `__nonexhaustive: ()` as a second field? That has no runtime impact, but lets people continue to use the field as public.

Luro02 replied:
I do not know if this would be good. Most people expect to obtain a `Span` with `Spanned::span`, but in this case the compiler would suggest to use the field instead.

TedDriggs replied:
Oh, that's a really good point.

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -150,17 +152,48 @@ impl<V: UsesLifetimes, F: UsesLifetimes> UsesLifetimes for Data<V, F> {
+        }
+    }
+
+    /// Adds a [`Span`] to [`Fields`].
```
How are these linking to something?

Luro02 replied:
There was an RFC, that added this feature
https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md#additions-to-the-documentation-syntax

TedDriggs replied:
That RFC hasn't been implemented yet: https://github.com/rust-lang/rust/issues/43466

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -219,52 +254,94 @@ impl<T> Fields<T> {
+
+impl<T: Eq> Eq for Fields<T> {}
+
+impl<F: FromField> Fields<F> {
```
Minor nit: I'd prefer to keep the inherent impl block above the trait impls. That should also reduce the diff size.

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -339,6 +417,7 @@ impl Style {
         Fields {
```
Should this use `Fields::new` so it's resilient to additional new fields?

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -219,52 +254,94 @@ impl<T> Fields<T> {
+        let fields = &self.fields;
+        // An unknown Span should be `Span::call_site()`;
+        // https://docs.rs/syn/1.0.12/syn/spanned/trait.Spanned.html#tymethod.span
+        let span = self.span.unwrap_or_else(|| Span::call_site());
```
Minor style question: Does this work?

```rust
let span = self.span.unwrap_or_else(Span::call_site);
```

Luro02 replied:
yes it does

TedDriggs commented on core/src/ast/data.rs:
```diff
@@ -150,18 +152,35 @@ impl<V: UsesLifetimes, F: UsesLifetimes> UsesLifetimes for Data<V, F> {
 pub struct Fields<T> {
     pub style: Style,
     pub fields: Vec<T>,
+    pub span: Option<Span>,
```
You've convinced me it's better for `Fields` to impl `Spanned` rather than exposing this as a field, but I don't see that trait impl anywhere. Am I missing it?

Luro02 replied:
`Spannend` is implemented for any type that implements `ToTokens`, which obtains the `Span` from the returned `TokenStream`.

WildCryptoFox commented:
FYI: This [was indeed a breaking change](https://github.com/sharksforarms/deku/pull/148).
